From 290f115b1dd613e29fc483f432a3243d35742d6d Mon Sep 17 00:00:00 2001
From: Hyun Kwon <hyun.kwon@xilinx.com>
Date: Thu, 26 Jan 2017 14:02:40 -0800
Subject: [PATCH 1234/1566] Revert "pll: zynqmp: Remove the disable for pll"

commit  242d025c94b9fc8f5ad5adb44de832e75332d19d from
https://github.com/Xilinx/linux-xlnx.git

With the current pmufw the disable of pll does not seem to work as
expected, ie disable once all the dependent clocks
are disabled. So revert the 18b0413295bae59f9db09aa459e27fd1ffbfa9b4
(pll disable) for now. Eventually once the pmufw
ref-counting happens correctly we can go remove the pll stuff from
linux.

Signed-off-by: Shubhrajyoti Datta <shubhrajyoti.datta@xilinx.com>
Signed-off-by: Hyun Kwon <hyun.kwon@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/clk/zynqmp/pll.c |   20 ++++++++++++++++++++
 1 files changed, 20 insertions(+), 0 deletions(-)

diff --git a/drivers/clk/zynqmp/pll.c b/drivers/clk/zynqmp/pll.c
index ea8767a..0c15a86 100644
--- a/drivers/clk/zynqmp/pll.c
+++ b/drivers/clk/zynqmp/pll.c
@@ -174,8 +174,28 @@ static int zynqmp_pll_enable(struct clk_hw *hw)
 	return 0;
 }
 
+/**
+ * zynqmp_pll_disable - Disable clock
+ * @hw:		Handle between common and hardware-specific interfaces
+ *
+ */
+static void zynqmp_pll_disable(struct clk_hw *hw)
+{
+	struct zynqmp_pll *clk = to_zynqmp_pll(hw);
+
+	if (!zynqmp_pll_is_enabled(hw))
+		return;
+
+	pr_info("PLL: shutdown\n");
+
+	/* shut down PLL */
+	zynqmp_pm_mmio_write((u32)(ulong)clk->pll_ctrl, PLLCTRL_RESET_MASK,
+				PLLCTRL_RESET_VAL);
+}
+
 static const struct clk_ops zynqmp_pll_ops = {
 	.enable = zynqmp_pll_enable,
+	.disable = zynqmp_pll_disable,
 	.is_enabled = zynqmp_pll_is_enabled,
 	.round_rate = zynqmp_pll_round_rate,
 	.recalc_rate = zynqmp_pll_recalc_rate
-- 
1.7.5.4

